File: Jarvis Visual/tests/test_why.py
#!/usr/bin/env python3
"""A DENY asks Serge why, and his answer reaches Jarvis.

Run:  ./tests/run-tests.sh   (from Jarvis Visual/)
   or  python3 tests/test_why.py

WHY THIS EXISTS, in his own words (2026-08-07 ~9:54 PM, after an evening of
denials that each ended with him being asked the same question by voice):

    "The deny did not work as it should. When I press the deny, there should
    be an inbox asking me why, and either me entering the reason or saying
    the reason with my voice. That's what I'm always waiting for."

And the rule underneath it, from the same conversation: "You have to know
why. If I don't say why, you just wait until I tell you why."

THE DEFECT THIS GUARDS IS NOT A MISSING FEATURE -- IT IS A HALF-WIRED ONE.
`clean_reason()` and `resolve_approval(..., reason)` already existed. What did
not exist was any path that CARRIED a reason: the HTTP route never read the
field, and `ask_permission` popped the request off `self.approvals` without
ever looking at what had been written to it. A reason could be stored and was
guaranteed to be thrown away one line later. So every test here drives the
real call chain end to end; none of them assert that a line of source exists.

That distinction is this project's oldest lesson, in its eighth costume: a
guard that proves a line EXISTS never proves it RUNS.
"""

import asyncio
import importlib.util
import json
import sys
import unittest
from pathlib import Path

HERE = Path(__file__).resolve().parent
SERVER = HERE.parent / "voice-web-server.py"


def load_server():
    spec = importlib.util.spec_from_file_location("voice_web_server", SERVER)
    mod = importlib.util.module_from_spec(spec)
    sys.modules["voice_web_server"] = mod
    spec.loader.exec_module(mod)
    return mod


srv = load_server()


def make_vw():
    """A VoiceWeb built without touching the real Brain."""
    vw = object.__new__(srv.VoiceWeb)
    vw.approvals = {}
    vw._approval_seq = 0
    vw.denied = None
    vw.reinstated = None
    return vw


class FakeRequest:
    """The one thing the routes touch: an awaitable .json()."""

    def __init__(self, body, raise_it=False):
        self._body = body
        self._raise = raise_it

    async def json(self):
        if self._raise:
            raise ValueError("not json")
        return self._body


def body_of(response):
    return json.loads(response.body.decode())


def run(coro):
    return asyncio.run(coro)


# ---------------------------------------------------------------------------
# THE WIRE: his reason has to survive the trip from the button to the request
# ---------------------------------------------------------------------------

class TestTheReasonReachesThePendingRequest(unittest.TestCase):
    """resolve_approval already stored it. Nothing ever PUT it there."""

    def setUp(self):
        self.vw = make_vw()
        srv.VW = self.vw
        loop = asyncio.new_event_loop()
        self.addCleanup(loop.close)
        self.fut = loop.create_future()
        self.vw.approvals[1] = {"id": 1, "tool": "Bash",
                                "detail": "ls", "future": self.fut}

    def test_the_route_carries_his_words_onto_the_request(self):
        # THE BUG, stated as a test: approval_reply used to call
        # resolve_approval with two arguments. The reason field on the POST
        # was read by nothing at all, so the box could send it forever and it
        # would land nowhere.
        r = run(srv.approval_reply(FakeRequest(
            {"id": 1, "allow": False, "reason": "wrong file"})))
        self.assertTrue(body_of(r)["ok"])
        self.assertEqual(self.vw.approvals[1]["reason"], "wrong file")

    def test_the_reason_is_cleaned_on_the_way_in(self):
        # Not re-testing clean_reason (test_denial.py owns that) -- testing
        # that this path actually PASSES THROUGH it rather than storing raw.
        run(srv.approval_reply(FakeRequest(
            {"id": 1, "allow": False, "reason": "  spaced  "})))
        self.assertEqual(self.vw.approvals[1]["reason"], "spaced")

    def test_no_reason_field_at_all_still_answers(self):
        # Every page older than tonight sends exactly this. A deny that failed
        # because it carried no explanation would break the one button on this
        # HUD that must always work.
        r = run(srv.approval_reply(FakeRequest({"id": 1, "allow": False})))
        self.assertTrue(body_of(r)["ok"])
        self.assertIsNone(self.vw.approvals[1]["reason"])
        self.assertFalse(self.fut.result())

    def test_a_reason_does_not_change_the_verdict(self):
        # The verdict is the future's result and the reason rides beside it.
        # If ever the two were merged, an explained deny could read as allow.
        run(srv.approval_reply(FakeRequest(
            {"id": 1, "allow": True, "reason": "go ahead"})))
        self.assertTrue(self.fut.result())

    def test_a_reason_on_an_unknown_id_grants_nothing(self):
        r = run(srv.approval_reply(FakeRequest(
            {"id": 99, "allow": True, "reason": "please"})))
        self.assertFalse(body_of(r)["ok"])
        self.assertFalse(self.fut.done())

    def test_a_junk_reason_is_dropped_not_fatal(self):
        # The field is free text from a browser; a number or a dict must not
        # take down the permission answer.
        for junk in (42, {"a": 1}, ["x"], True):
            with self.subTest(junk=junk):
                vw = make_vw()
                srv.VW = vw
                loop = asyncio.new_event_loop()
                self.addCleanup(loop.close)
                fut = loop.create_future()
                vw.approvals[1] = {"id": 1, "tool": "Bash",
                                   "detail": "ls", "future": fut}
                r = run(srv.approval_reply(FakeRequest(
                    {"id": 1, "allow": False, "reason": junk})))
                self.assertTrue(body_of(r)["ok"])
                self.assertIsNone(vw.approvals[1]["reason"])


# ---------------------------------------------------------------------------
# THE HAND-OFF: ask_permission must READ it before discarding the request
# ---------------------------------------------------------------------------

class TestTheDenialMessageCarriesHisReason(unittest.IsolatedAsyncioTestCase):
    """The half that was structurally impossible before this change.

    ask_permission pops the request in its `finally`. Whatever
    resolve_approval wrote is on that popped dict and nowhere else, so
    reading self.approvals afterwards can only ever find nothing.
    """

    async def _deny_with(self, reason):
        vw = make_vw()
        srv.VW = vw
        srv.WS_CLIENTS = set()

        async def answer():
            # Wait until the request is really registered, then answer it the
            # way the button does -- through the route, not by hand.
            for _ in range(200):
                if vw.approvals:
                    break
                await asyncio.sleep(0.001)
            aid = next(iter(vw.approvals))
            body = {"id": aid, "allow": False}
            if reason is not None:
                body["reason"] = reason
            await srv.approval_reply(FakeRequest(body))

        task = asyncio.create_task(answer())
        result = await vw.ask_permission("Bash", {"command": "ls"}, None)
        await task
        return vw, result

    async def test_his_words_are_in_the_message_handed_to_the_brain(self):
        vw, result = await self._deny_with("that file is not yours to touch")
        self.assertIn("that file is not yours to touch", result.message)

    async def test_a_given_reason_forbids_asking_him_again(self):
        # THE WHOLE POINT, and the thing he was angry about tonight: he
        # answers, and is asked the same question back.
        _, result = await self._deny_with("wrong branch")
        self.assertIn("Do NOT ask him why", result.message)

    async def test_no_reason_means_ask_and_then_wait(self):
        # His rule: "If I don't say why, you just wait until I tell you why."
        _, result = await self._deny_with(None)
        self.assertIn("Ask him why", result.message)
        self.assertIn("WAIT", result.message)
        self.assertNotIn("Do NOT ask him why", result.message)

    async def test_the_two_cases_are_genuinely_different_messages(self):
        _, with_why = await self._deny_with("because")
        _, without = await self._deny_with(None)
        self.assertNotEqual(with_why.message, without.message)

    async def test_his_words_are_delimited_rather_than_loose_in_the_prompt(self):
        # His reason is free text arriving from a page field. It is his own
        # voice and it is allowed to steer me -- but it must be legible AS a
        # quotation, not blended into the server's own framing.
        _, result = await self._deny_with("stop and do the tests first")
        self.assertIn("<<<stop and do the tests first>>>", result.message)

    async def test_the_refusal_record_keeps_the_reason(self):
        # The red card and the one-use pass hang off vw.denied. A reason that
        # lived only in the message would be gone the moment I answered him.
        vw, _ = await self._deny_with("not now")
        self.assertEqual(vw.denied["reason"], "not now")

    async def test_a_denial_with_no_reason_records_None_not_the_string(self):
        vw, _ = await self._deny_with(None)
        self.assertIsNone(vw.denied["reason"])

    async def test_the_stop_instruction_survives_in_both_cases(self):
        # The reason is an ADDITION. If explaining a refusal quietly dropped
        # "STOP HERE", an explained deny would be weaker than a silent one.
        for reason in ("because", None):
            with self.subTest(reason=reason):
                _, result = await self._deny_with(reason)
                self.assertIn("STOP HERE", result.message)

    async def test_an_approval_is_unaffected_by_the_new_field(self):
        vw = make_vw()
        srv.VW = vw
        srv.WS_CLIENTS = set()

        async def answer():
            for _ in range(200):
                if vw.approvals:
                    break
                await asyncio.sleep(0.001)
            aid = next(iter(vw.approvals))
            await srv.approval_reply(FakeRequest(
                {"id": aid, "allow": True, "reason": "fine by me"}))

        task = asyncio.create_task(answer())
        result = await vw.ask_permission("Bash", {"command": "ls"}, None)
        await task
        self.assertIsInstance(result, srv.PermissionResultAllow)
        self.assertIsNone(vw.denied)


# ---------------------------------------------------------------------------
# SPEAKING THE REASON -- transcription that is ONLY transcription
# ---------------------------------------------------------------------------

class TestTheSpokenReasonRoute(unittest.IsolatedAsyncioTestCase):
    """It turns his voice into text for the BOX. It must do nothing else.

    While a permission is pending, a second path into the brain is the one
    thing that must not exist -- so this route's safety is that it has none.
    """

    def setUp(self):
        self.vw = make_vw()
        self.heard = []

        async def fake_transcribe(wav):
            self.heard.append(wav)
            return "because the tests are red"

        self.vw.transcribe = fake_transcribe
        srv.VW = self.vw

    @staticmethod
    def _wav(seconds=0.1):
        import base64
        import io
        import wave
        buf = io.BytesIO()
        with wave.open(buf, "wb") as w:
            w.setnchannels(1)
            w.setsampwidth(2)
            w.setframerate(16000)
            w.writeframes(b"\x00\x01" * int(16000 * seconds))
        return base64.b64encode(buf.getvalue()).decode()

    async def test_it_gives_back_what_he_said(self):
        r = await srv.reason_transcribe(FakeRequest({"audio": self._wav()}))
        d = body_of(r)
        self.assertTrue(d["ok"])
        self.assertEqual(d["text"], "because the tests are red")

    async def test_it_never_touches_the_brain_or_an_approval(self):
        # Proven by state, not by reading the function: a pending request must
        # be exactly as pending afterwards, and no denial may appear.
        loop = asyncio.get_running_loop()
        self.vw.approvals[1] = {"id": 1, "tool": "Bash", "detail": "ls",
                                "future": loop.create_future()}
        await srv.reason_transcribe(FakeRequest({"audio": self._wav()}))
        self.assertFalse(self.vw.approvals[1]["future"].done())
        self.assertIsNone(self.vw.denied)
        self.assertIsNone(self.vw.reinstated)

    async def test_no_audio_is_refused_in_words_he_can_read(self):
        for bad in (None, "", 5, {"a": 1}, []):
            with self.subTest(bad=bad):
                r = await srv.reason_transcribe(FakeRequest({"audio": bad}))
                d = body_of(r)
                self.assertFalse(d["ok"])
                self.assertTrue(d["error"])

    async def test_a_body_that_is_not_json_is_refused_not_fatal(self):
        r = await srv.reason_transcribe(FakeRequest(None, raise_it=True))
        self.assertFalse(body_of(r)["ok"])

    async def test_unreadable_base64_is_refused(self):
        r = await srv.reason_transcribe(FakeRequest({"audio": "!!!not b64!!!"}))
        d = body_of(r)
        self.assertFalse(d["ok"])
        self.assertEqual(self.heard, [])      # never reached whisper

    async def test_an_oversized_post_is_refused_before_it_is_decoded(self):
        # The only caller-supplied blob on this server. Cheaper to refuse than
        # to decode, and the decode never happens.
        #
        # ⚠ THE FIRST VERSION OF THIS TEST PASSED WITH THE CAP DELETED, and
        # the injection round is what found it. It padded with "A" to an odd
        # length, so base64 rejected it as malformed and the route said
        # ok:false for a reason that had nothing to do with size -- the test
        # could not tell "too long" from "unreadable". The payload is now
        # VALID base64, so without the cap it would decode and travel on, and
        # the assertion names the refusal it expects rather than any refusal.
        oversize = "A" * (((srv.MAX_REASON_WAV_B // 4) + 1) * 4)
        self.assertGreater(len(oversize), srv.MAX_REASON_WAV_B)
        r = await srv.reason_transcribe(FakeRequest({"audio": oversize}))
        d = body_of(r)
        self.assertFalse(d["ok"])
        self.assertIn("too long", d["error"])
        self.assertEqual(self.heard, [])

    async def test_a_whisper_failure_answers_rather_than_raising(self):
        # If this raised, aiohttp would answer 500 and the page would show him
        # nothing -- and he would be left with a box that appears broken and
        # no hint that he can simply type instead.
        async def boom(wav):
            raise RuntimeError("whisper is down")

        self.vw.transcribe = boom
        r = await srv.reason_transcribe(FakeRequest({"audio": self._wav()}))
        d = body_of(r)
        self.assertFalse(d["ok"])
        self.assertIn("type it", d["error"])

    async def test_the_route_is_actually_registered(self):
        # A route that exists as a function and is never mounted is a 404 to
        # him, which on this page looks exactly like the feature being broken.
        src = SERVER.read_text()
        self.assertIn('web.post("/reason-transcribe", reason_transcribe)', src)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: Jarvis Visual/tests/test_why.js
#!/usr/bin/env node
// The why-box on the JARVIS page.
//
// Serge, 2026-08-07 ~9:54 PM: "When I press the deny, there should be an
// inbox asking me why, and either me entering the reason or saying the reason
// with my voice. That's what I'm always waiting for."
//
// As with the other page tests, the real functions are pulled out of
// jarvis.html and RUN against stubs, so they cannot drift from what ships --
// and so that none of these can pass by matching a line that never executes,
// which is the failure this project keeps rediscovering.

const fs = require('fs');
const path = require('path');
const assert = require('assert');

const HTML = path.join(__dirname, '..', 'jarvis.html');
const src = fs.readFileSync(HTML, 'utf8');

function grab(name) {
  const re = new RegExp('(?:^|\\n)(?:async )?function ' + name + '\\s*\\([^)]*\\)\\s*\\{');
  const m = src.match(re);
  if (!m) throw new Error('function ' + name + ' not found in jarvis.html');
  const start = src.indexOf(m[0]) + (m[0].startsWith('\n') ? 1 : 0);
  let i = src.indexOf('{', start), depth = 0;
  for (let j = i; j < src.length; j++) {
    if (src[j] === '{') depth++;
    else if (src[j] === '}') { depth--; if (!depth) return src.slice(start, j + 1); }
  }
  throw new Error('unbalanced braces reading ' + name);
}

// Code only. A comment that NAMES the thing it forbids is the single most
// repeated way a guard on this project has passed against broken code.
function noComments(s) {
  return s.replace(/\/\*[\s\S]*?\*\//g, '').replace(/(^|\n)\s*\/\/[^\n]*/g, '$1');
}

// ⚠ THE RUNNER AWAITS. The first version of this file called fn() and moved
// on, so every `async () => {...}` test returned a promise nobody looked at:
// its assertions ran after the summary, and a REJECTION printed "ok". Three
// of these tests were passing against code that threw. A test harness that
// cannot fail is the same defect as a guard that cannot fire.
let tests = 0, failed = 0;
const queue = [];
function t(name, fn) {
  tests++;
  queue.push(async () => {
    try { await fn(); console.log('  ok   ' + name); }
    catch (e) { failed++; console.log('  FAIL ' + name + '\n       ' + e.message); }
  });
}
function section(title) { queue.push(async () => console.log('\n=== ' + title)); }

// ---- DOM stub -------------------------------------------------------------
function el() {
  return { style: {display: ''}, value: '', textContent: '', focused: false,
           classes: new Set(),
           classList: { add(c){ this.owner.classes.add(c); },
                        remove(c){ this.owner.classes.delete(c); } },
           focus(){ this.focused = true; },
           addEventListener(){}, };
}
function mkEl() { const e = el(); e.classList.owner = e; return e; }

// BUILT ONCE AND ONLY CLEARED, never replaced. The page's why-box block
// captures its elements in `const`s at load, exactly as the browser does, so
// handing it a fresh set of nodes between tests would leave it holding the
// old ones -- and every test would run against a DOM nothing could see.
const nodes = {};
for (const id of ['why-wrap', 'why-text', 'why-note', 'why-speak',
                  'why-send', 'why-skip', 'approve-yes', 'approve-no',
                  'approve-head', 'approve-detail', 'approve-box'])
  nodes[id] = mkEl();
function resetDom() {
  for (const id of Object.keys(nodes)) {
    const e = nodes[id];
    e.style.display = ''; e.value = ''; e.textContent = '';
    e.focused = false; e.classes.clear();
  }
}
const window = { addEventListener(){} };
const document = { getElementById: (id) => nodes[id] || null, body: { classList: { add(){}, remove(){} } } };

// ---- state the extracted functions close over -----------------------------
let approvalId = null, approvalQueued = null, approvalSending = false,
    approvalStale = false, approvalHttp = true, serverBoot = 'b1',
    capturing = false, approveBox = null, statusSaid = '';
// whyCapturing / whyBufs / whyBusy are declared by the page block itself and
// `let` inside eval() lives in the EVAL's scope -- so declaring them out here
// would silently shadow nothing and leave every poke invisible to the code
// under test. That is exactly how the transcription-failure test first
// "passed": whyCapturing was false in the real function, so it returned on
// its first line and asserted against a box nothing had touched.
const MIN_CAP_S = 0.20;
let micCtx = {sampleRate: 16000, state: 'running', resume(){}};
let posts = [];
const REPLY_TIMEOUT_MS = 5000;

function approvalStatus(s) { statusSaid = s; }
class AbortController { constructor(){ this.signal = {}; } abort(){} }
const setTimeout_ = setTimeout, clearTimeout_ = clearTimeout;

let fetchResult = {ok: true, json: async () => ({ok: true})};
async function fetch(url, opts) {
  posts.push({url, body: JSON.parse(opts.body)});
  if (fetchResult instanceof Error) throw fetchResult;
  return fetchResult;
}

// The why-box block is a run of top-level declarations, not one function, so
// it is sliced by its own banner rather than by grab().
const WHY_START = src.indexOf('const whyWrap  = document.getElementById');
const WHY_END = src.indexOf("document.getElementById('approve-yes').onclick");
assert.ok(WHY_START > 0 && WHY_END > WHY_START, 'why-box block not found in jarvis.html');
const whyBlock = src.slice(WHY_START, WHY_END);

eval(whyBlock + '\n'
     + grab('retryWords') + '\n'
     + grab('deliverApproval') + '\n'
     + grab('answerApproval') + '\n'
     + grab('flushApproval') + '\n'
     + grab('showApproval') + '\n'
     // Evaluated in the SAME scope as the block, which is the only way to
     // reach its `let`s. A test that cannot set the state it is testing is
     // not a test.
     + 'function __why(cap, bufs, busy){ whyCapturing = cap; whyBufs = bufs || [];'
     + ' whyBusy = !!busy; }');

function reset() {
  resetDom();
  approveBox = nodes['approve-box'];
  approvalId = 1; approvalQueued = null; approvalSending = false;
  approvalStale = false; approvalHttp = true; serverBoot = 'b1';
  __why(false, [], false); capturing = false;
  posts = []; statusSaid = '';
  fetchResult = {ok: true, json: async () => ({ok: true})};
}

// Re-eval inside reset() rebinds consts in this scope; simpler is to keep the
// handles as lookups. Redefine the four helpers against live lookups instead.
const whyWrapOf  = () => nodes['why-wrap'];
const whyTextOf  = () => nodes['why-text'];
const whyNoteOf  = () => nodes['why-note'];

section('the box opens on DENY, and DENY alone does not send');

t('DENY opens the box and sends nothing', () => {
  reset();
  openWhy();
  assert.strictEqual(whyWrapOf().style.display, 'block');
  assert.strictEqual(posts.length, 0, 'a deny was sent before he said why');
});

t('the box does not open when there is no request to refuse', () => {
  reset();
  approvalId = null;
  openWhy();
  assert.notStrictEqual(whyWrapOf().style.display, 'block');
});

t('the textarea takes focus so he can just type', () => {
  reset();
  openWhy();
  assert.ok(whyTextOf().focused);
});

section('SEND carries his words with the refusal, in ONE post');

t('the reason and the verdict travel together', async () => {
  reset();
  openWhy();
  whyTextOf().value = 'that is the wrong branch';
  sendWhy();
  await new Promise(r => setImmediate(r));
  assert.strictEqual(posts.length, 1, 'expected exactly one post');
  assert.strictEqual(posts[0].url, '/approval-reply');
  assert.strictEqual(posts[0].body.allow, false);
  assert.strictEqual(posts[0].body.reason, 'that is the wrong branch');
});

t('the reason is trimmed before it is sent', async () => {
  reset();
  openWhy();
  whyTextOf().value = '   spaced out   ';
  sendWhy();
  await new Promise(r => setImmediate(r));
  assert.strictEqual(posts[0].body.reason, 'spaced out');
});

t('an empty SEND says so instead of doing nothing', async () => {
  reset();
  openWhy();
  whyTextOf().value = '   ';
  sendWhy();
  await new Promise(r => setImmediate(r));
  assert.strictEqual(posts.length, 0, 'an empty reason was sent');
  assert.ok(whyNoteOf().textContent.length > 0,
            'the press was silently ignored');
});

t('APPROVE sends no reason at all', async () => {
  reset();
  answerApproval(true);
  await new Promise(r => setImmediate(r));
  assert.strictEqual(posts[0].body.allow, true);
  assert.strictEqual(posts[0].body.reason, null);
});

section('he is never trapped in the box');

t('DENY WITHOUT A REASON sends the refusal with a null reason', async () => {
  reset();
  openWhy();
  whyTextOf().value = 'half a thought';
  answerApproval(false, null);
  await new Promise(r => setImmediate(r));
  assert.strictEqual(posts.length, 1);
  assert.strictEqual(posts[0].body.allow, false);
  assert.strictEqual(posts[0].body.reason, null,
                     'skipping still leaked the half-typed text');
});

section('a reason never bleeds onto a request it was not written for');

t('sending clears the box', async () => {
  reset();
  openWhy();
  whyTextOf().value = 'first reason';
  sendWhy();
  await new Promise(r => setImmediate(r));
  assert.strictEqual(whyTextOf().value, '');
  assert.notStrictEqual(whyWrapOf().style.display, 'block');
});

t('a NEW request closes the box and drops the old text', () => {
  reset();
  openWhy();
  whyTextOf().value = 'about the OLD request';
  showApproval({id: 2, tool: 'Bash', detail: 'rm -rf'});
  assert.strictEqual(whyTextOf().value, '',
                     "the previous request's reason survived into a new one");
  assert.notStrictEqual(whyWrapOf().style.display, 'block');
});

t('the request going away closes the box', () => {
  reset();
  openWhy();
  whyTextOf().value = 'typing…';
  showApproval(null);
  assert.strictEqual(whyTextOf().value, '');
  assert.notStrictEqual(whyWrapOf().style.display, 'block');
});

t('a REDRAW of the same request leaves what he is typing alone', () => {
  reset();
  openWhy();
  whyTextOf().value = 'half a sentence';
  showApproval({id: 1, tool: 'Bash', detail: 'ls'});   // the 15Hz poll
  assert.strictEqual(whyTextOf().value, 'half a sentence',
                     'the poll ate his words mid-typing');
});

section('his words survive a failed send');

t('a refused post keeps the reason queued with the verdict', async () => {
  reset();
  fetchResult = {ok: false, json: async () => ({ok: false, error: 'nope'})};
  openWhy();
  whyTextOf().value = 'because I said so';
  sendWhy();
  await new Promise(r => setImmediate(r));
  assert.ok(approvalQueued, 'his click was dropped');
  assert.strictEqual(approvalQueued.reason, 'because I said so',
                     'the retry would deliver an unexplained refusal');
});

t('the queued retry delivers the reason too', async () => {
  reset();
  approvalQueued = {id: 1, allow: false, reason: 'kept', boot: 'b1'};
  await flushApproval();
  assert.strictEqual(posts[0].body.reason, 'kept');
});

t('a superseding click keeps its own reason', async () => {
  reset();
  approvalSending = true;
  answerApproval(false, 'the second thought');
  assert.strictEqual(approvalQueued.reason, 'the second thought');
});

section('speaking the reason is transcription and nothing else');

t('the spoken buffer is never the turn buffer', () => {
  // capBufs is what becomes a spoken TURN. If the why-box wrote into it, a
  // refusal reason could be sent to the brain as speech while a permission
  // is pending -- the one thing that must not happen.
  // COMMENTS STRIPPED FIRST, and that is not a detail: the first version of
  // this test read the raw slice and failed against correct code, because the
  // comment explaining why the buffers are separate says "capBufs". Seventh
  // time on this project's record that grepping source punished the prose
  // explaining the decision -- walked into again while writing the guard.
  const branch = noComments(
    grab('ensureMic').slice(grab('ensureMic').indexOf('else if (whyCapturing)')));
  assert.ok(branch.includes('whyBufs.push'), 'the why branch fills no buffer');
  assert.ok(!branch.includes('capBufs'),
            'the why-box writes into the turn buffer');
});

t('the talk button stands down while he is speaking a reason', () => {
  const press = grab('press');
  assert.ok(/if \(whyCapturing \|\| whyBusy\) return;/.test(press),
            'press() would grab the microphone mid-reason');
});

t('the transcription post goes to the transcribe-only route', () => {
  assert.ok(src.includes("fetch('/reason-transcribe'"),
            'the spoken reason has no route to text');
  // And it must NOT be the socket, which is how a turn is opened.
  const stop = grab('whyStopSpeaking');
  assert.ok(!/ws\.send/.test(stop),
            'the spoken reason can reach the brain as a turn');
});

t('a transcription failure still leaves him able to type', async () => {
  reset();
  fetchResult = new Error('network down');
  __why(true, [new Float32Array(8000)]);
  await whyStopSpeaking(false);
  assert.ok(/type it/i.test(whyNoteOf().textContent),
            'a failed transcription said nothing about typing');
});

t('heard text is appended, never overwriting what he already typed', () => {
  const stop = grab('whyStopSpeaking');
  assert.ok(/whyText\.value \? \(whyText\.value\.trim\(\) \+ ' ' \+ heard\) : heard/.test(stop),
            'transcription replaces rather than appends');
});

section('the wiring actually points at the box');

t('the DENY button opens the box rather than sending a refusal', () => {
  assert.ok(/getElementById\('approve-no'\)\.onclick = openWhy;/.test(src),
            'DENY does not open the why box');
  assert.ok(!/getElementById\('approve-no'\)\.onclick = \(\) => answerApproval\(false\)/.test(src),
            'DENY still sends the old unexplained refusal');
});

t('both ways out of the box are wired', () => {
  assert.ok(/getElementById\('why-send'\)\.onclick = sendWhy;/.test(src));
  assert.ok(/getElementById\('why-skip'\)\.onclick = \(\) => answerApproval\(false, null\);/.test(src));
});

t('a bare Enter does not send his reason', () => {
  // This page has already cost him one popup to a stray Enter key.
  const m = src.match(/whyText\.addEventListener\('keydown'[\s\S]{0,200}?\}\);/);
  assert.ok(m, 'no keydown handler on the reason field');
  assert.ok(/metaKey \|\| e\.ctrlKey/.test(m[0]),
            'Enter alone would send an unfinished reason');
});

t('the box ships CLOSED', () => {
  assert.ok(/#why-wrap \{ display: none;/.test(src),
            'the why box would show on every permission request');
});

section('the talk key belongs to the box while the box is open');

// Serge, 2026-08-07 ~10:50 PM: "Right command did not work, but if I press
// the button, then I speak, then the text went in the input." It was not
// merely unwired -- the key still drove press()/release(), so his spoken
// reason opened a TURN and arrived as a message to Jarvis. These run the real
// handlers rather than reading them: the fault was invisible to the eye and
// the last thing this file needs is another guard that only greps.
function keyHandler(kind) {
  // Slice the MetaRight handler of the given kind out of the page. Anchored
  // on the code that actually distinguishes them, so a reordering of the two
  // cannot silently hand a test the wrong one.
  const re = new RegExp("addEventListener\\('" + kind + "', \\(e\\) => \\{[\\s\\S]*?\\n\\}\\);");
  const all = src.match(new RegExp(re, 'g')) || [];
  const one = all.filter(s => /MetaRight/.test(s));
  assert.strictEqual(one.length, 1,
    'expected exactly one MetaRight ' + kind + ' handler, found ' + one.length);
  const body = one[0].replace(/^addEventListener\('[a-z]+', \(e\) => \{/, '').replace(/\}\);$/, '');
  assert.ok(/whyIsOpen|whyCapturing/.test(noComments(body)),
            'the ' + kind + ' handler no longer consults the why box');
  return new Function(
    'e', 'lineAlive', 'typedEl', 'document', 'whyIsOpen', 'whyStartSpeaking',
    'press', 'whyCapturing', 'whyStopSpeaking', 'release', body);
}

function keyRun(kind, {open = false, cap = false, focus = 'body', line = false} = {}) {
  const calls = [];
  const typed = {};
  const active = focus === 'typed' ? typed : {};
  const e = {code: 'MetaRight', repeat: false, prevented: false,
             preventDefault(){ this.prevented = true; }};
  keyHandler(kind)(
    e, line, typed, {activeElement: active},
    () => open,
    () => calls.push('whyStart'),
    () => calls.push('press'),
    cap,
    (d) => calls.push('whyStop:' + d),
    () => calls.push('release'));
  return {calls, e};
}

t('holding the key with the box open speaks INTO the box', () => {
  const r = keyRun('keydown', {open: true});
  assert.deepStrictEqual(r.calls, ['whyStart'],
    'his refusal reason opened a spoken turn instead of filling the box');
});

t('holding the key with the box shut still talks to Jarvis', () => {
  const r = keyRun('keydown', {open: false});
  assert.deepStrictEqual(r.calls, ['press'],
    'the ordinary talk key was broken by the why-box guard');
});

t('releasing while the box is capturing ENDS the box capture, and keeps it', () => {
  const r = keyRun('keyup', {cap: true});
  assert.deepStrictEqual(r.calls, ['whyStop:false'],
    'the spoken reason was discarded or sent as a turn');
});

t('releasing with no box capture falls through to the talk button', () => {
  const r = keyRun('keyup', {cap: false});
  assert.deepStrictEqual(r.calls, ['release']);
});

t('a box CLOSED mid-hold still stops the microphone', () => {
  // The key-up asks whyCapturing, never whyIsOpen -- closing the box between
  // press and release must not strand the mic live with no way back.
  const r = keyRun('keyup', {open: false, cap: true});
  assert.deepStrictEqual(r.calls, ['whyStop:false'],
    'the microphone was left capturing after the box closed');
});

t('the terminal line still owns the key, box or no box', () => {
  const r = keyRun('keydown', {open: true, line: true});
  assert.deepStrictEqual(r.calls, [],
    'the page grabbed Right Command while the terminal line held it');
});

t('typing in the command box is never a talk key', () => {
  assert.deepStrictEqual(keyRun('keydown', {focus: 'typed', open: true}).calls, []);
  assert.deepStrictEqual(keyRun('keyup', {focus: 'typed', cap: true}).calls, []);
});

t('a key REPEAT does not restart the box capture', () => {
  const calls = [];
  const typed = {};
  keyHandler('keydown')(
    {code: 'MetaRight', repeat: true, preventDefault(){}},
    false, typed, {activeElement: {}},
    () => true, () => calls.push('whyStart'), () => calls.push('press'),
    false, () => {}, () => {});
  assert.deepStrictEqual(calls, [], 'holding the key re-armed the capture');
});

t('the talk path and the box path are mutually exclusive', () => {
  // Both firing would send the audio twice -- once as a turn, once as text.
  for (const open of [true, false]) {
    const r = keyRun('keydown', {open});
    assert.strictEqual(r.calls.length, 1, 'both capture paths ran on one press');
  }
});

// ⚠ THE ONE FAULT THAT RAN UNCAUGHT: `whyIsOpen(){ return false; }` left the
// whole suite green, because every handler test above STUBS it. The key tests
// prove the routing; only these prove the thing the routing asks. A guard the
// tests supply for themselves is not a guard.
t('whyIsOpen answers the real box, opened the real way', () => {
  reset();
  assert.strictEqual(whyIsOpen(), false, 'the box reads open before any DENY');
  openWhy();
  assert.strictEqual(whyIsOpen(), true, 'a DENY leaves the key talking to Jarvis');
});

t('whyIsOpen goes false again when the box closes', () => {
  reset();
  openWhy();
  closeWhy();
  assert.strictEqual(whyIsOpen(), false,
    'the key would keep feeding a box that is no longer on screen');
});

t('whyIsOpen is not a constant', () => {
  // Kills the degenerate `return true` / `return false` bodies outright.
  reset();
  const shut = whyIsOpen();
  openWhy();
  assert.notStrictEqual(whyIsOpen(), shut, 'whyIsOpen never changes its answer');
});

t('press() still stands down while the box holds the microphone', () => {
  // The guard inside press() is the second line of defence and must stay.
  assert.ok(/if \(whyCapturing \|\| whyBusy\) return;/.test(noComments(grab('press'))),
            'press() would open a turn over a reason being spoken');
});

(async () => {
  for (const step of queue) await step();
  console.log('\n' + (tests - failed) + '/' + tests + ' passed');
  if (failed) process.exit(1);
})();
